rectools/models/sasrec.py

```diff
@@ -561,13 +540,57 @@ def get_dataloader_recommend(self, dataset: Dataset) -> DataLoader:
         return recommend_dataloader
 
 
+# ####  --------------  Lightning Model  --------------  #### #
+
+
+class SessionEncoderLightningModule(SessionEncoderLightningModuleBase):
+    """TODO"""
+
+    def on_train_start(self) -> None:
+        """TODO"""
+        self._xavier_normal_init()
+
+    def training_step(self, batch: torch.Tensor, batch_idx: int) -> torch.Tensor:
+        """TODO"""
+        x, y, w = batch
+        logits = self.forward(x)  # [batch_size, session_maxlen, n_items + 1]
+        # We are using CrossEntropyLoss with a multi-dimensional case
+
+        # Logits must be passed in form of [batch_size, n_items + 1, session_maxlen],
+        #  where n_items + 1 is number of classes
+
+        # Target label indexes must be passed in a form of [batch_size, session_maxlen]
+        # (`0` index for "PAD" ix excluded from loss)
+
+        # Loss output will have a shape of [batch_size, session_maxlen]
+        # and will have zeros for every `0` target label
+        if self.loss == "softmax":
+            # ce_criterion = torch.nn.CrossEntropyLoss()
+            # https://github.com/NVIDIA/pix2pixHD/issues/9 how could an old bug appear again...
+            loss_func = nn.CrossEntropyLoss(ignore_index=0, reduction="none")
+            loss = loss_func(logits.transpose(1, 2), y)  # [batch_size, session_maxlen]
+            loss = loss * w
+            n = (loss > 0).to(loss.dtype)
+            loss = torch.sum(loss) / torch.sum(n)
+            return loss
+        raise ValueError(f"loss {loss} is not supported")
+
+    def _xavier_normal_init(self) -> None:
+        """TODO"""
+        for _, param in self.torch_model.named_parameters():
+            try:
+                torch.nn.init.xavier_normal_(param.data)
+            except ValueError:
+                pass
+
+
 # ####  --------------  SASRec Model  --------------  #### #
 
 
-class SasRecModel(ModelBase):  # pylint: disable=too-many-instance-attributes
+class SasRecModel(ModelBase):
     """TODO"""
 
-    def __init__(
+    def __init__(  # pylint: disable=too-many-arguments
```

Comment:
Let's add `dataloader_num_workers` argument and pass it to torch.Dataloader on initialization. This is DSSM example: https://github.com/MobileTeleSystems/RecTools/blob/8a3b7165a6928507a4edfe01ecd100280b7a0a00/rectools/models/dssm.py#L287
We missed this one but it is important.

Reply:
fixed